3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala

```diff
@@ -72,233 +74,288 @@ object EnrichmentManager {
-        } ++ jsScript.getOrElse(Nil) ++
-          cookieExtractorContexts ++
-          httpHeaderExtractorContexts ++
-          yauaaContext
-      } yield res
-
-    // Derive some contexts with custom SQL Query enrichment
-    val sqlQueryContexts: F[EnrichContextsComplex] =
-      getSqlQueryContexts(
-        event,
-        preparedDerivedContexts,
-        customContexts,
-        unstructEvent,
-        registry.sqlQuery
-      )
+        // Derive some contexts with custom API Request enrichment
+        val apiRequestContexts
+          : F[ValidatedNel[FailureDetails.EnrichmentFailure, List[SelfDescribingData[Json]]]] =
+          getApiRequestContexts[F](
+            event,
+            builtInContexts,
+            inputContexts,
+            unstructEvent,
+            registry.apiRequest
+          )
 
-    // Derive some contexts with custom API Request enrichment
-    val apiRequestContexts: F[EnrichContextsComplex] =
-      getApiRequestContexts(
-        event,
-        preparedDerivedContexts,
-        customContexts,
-        unstructEvent,
-        registry.apiRequest
-      )
+        // Merge all enrichments errors if any
+        (
+          currency,
+          weatherContext,
+          geoloc,
+          sqlQueryContexts,
+          apiRequestContexts,
+          builtInContexts
+        ).mapN { (cur, wea, _, sql, api, ctxts) =>
+          (
+            cur.toValidated,
+            wea.toValidated,
+            sql,
+            api,
+            noMonadEnrichments
+          ).mapN { (_, _, sqlContexts, apiContexts, _) =>
+              ctxts ++ sqlContexts ++ apiContexts
+            }
+            .leftMap(
+              enrichmentFailures =>
+                buildEnrichmentFailuresBadRow(
+                  enrichmentFailures,
+                  EnrichedEvent.toPartiallyEnrichedEvent(event),
+                  RawEvent.toRawEvent(raw),
+                  processor
+                )
+            )
+            .toEither
+        }
+      }
 
-    // Assemble prepared derived contexts with fetched via API Request
-    val derivedContexts: F[List[SelfDescribingData[Json]]] =
-      for {
-        api <- apiRequestContexts
-        sql <- sqlQueryContexts
-        prepDerivedContexts <- preparedDerivedContexts
-      } yield api.getOrElse(Nil) ++ sql.getOrElse(Nil) ++ prepDerivedContexts
+      //////////////////////////////////////////////////////////////////////
+      // 5. Check that all the contexts created by the enrichments are valid
+      //////////////////////////////////////////////////////////////////////
+
+      _ <- EitherT(for {
+        valid <- IgluUtils.validateEnrichmentsContexts[F](client, enrichmentsContexts)
+      } yield valid.leftMap { enrichmentFailures =>
+        buildEnrichmentFailuresBadRow(
+          enrichmentFailures,
+          EnrichedEvent.toPartiallyEnrichedEvent(event),
+          RawEvent.toRawEvent(raw),
+          processor
+        )
+      }.toEither)
```

Comment:
```suggestion
      _ <- EitherT(IgluUtils.validateEnrichmentsContexts[F](client, enrichmentsContexts).map(_.toEither))
        .leftMap { enrichmentFailures =>
          buildEnrichmentFailuresBadRow(
            enrichmentFailures,
            EnrichedEvent.toPartiallyEnrichedEvent(event),
            RawEvent.toRawEvent(raw),
            processor
          )
        }
```

Just to get rid of `for`-comprehension.